### PR TITLE
Laboratory work #4, Marina Mokina - 22FPL1

### PR DESCRIPTION
**Pull request by mmarina2004**

### CODE REVIEW

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -56,6 +75,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
...
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            raise ValueError
+
+        words_list = list(decoded_corpus)
+        sentences = (' '.join(words_list)).split('<eos>')
+        decoded_text = ''
+        for i, sentence in enumerate(sentences):
+            sentence = sentence.strip().capitalize()
+            decoded_text += f'{sentence}. '
+        if decoded_corpus[-1] == '<eos>':
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:87]: what is `<eos>`?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py:87]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -98,6 +134,35 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if not (isinstance(seq_len, int) and seq_len > 0
+                and isinstance(prompt, str) and prompt):
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -98,6 +134,35 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
+        for i in range(seq_len):
+            next_tokens = self._model.generate_next_token(encoded)
+            if next_tokens is None:
+                raise ValueError
+            if not next_tokens:
+                break
+
+            sorted_dict = dict(sorted(list(next_tokens.items()), key=lambda x: (x[1], x[0]), reverse=True))
```
[the remark this answers is not in the document]
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -28,6 +32,19 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError('WordProcessor._tokenize: Incorrect input')
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:36]: remove `WordProcessor._tokenize` from the message

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -28,6 +32,19 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+
+        tokens = []
+        for word in text.lower().split():
+            if word[-1] in '!?.':
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:40]: `'!?.'` to a variable

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -125,6 +195,8 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        generators = ['Greedy Generator', 'Top-P Generator', 'Beam Search Generator']
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:198]: what if `generator_type` is 4?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 4 above)
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 5 above)
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 6 above)
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -98,6 +136,36 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+                or seq_len <= 0):
+            raise ValueError("Incorrect input")
+        encoded = self._word_processor.encode(prompt)
+        if not encoded:
+            raise ValueError("Encoded is None")
+
+        for i in range(seq_len):
+            next_tokens = self._model.generate_next_token(encoded)
+            if next_tokens is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:149]: message

File: lab_4_fill_words_by_ngrams/main.py [hunk 8]
```diff
@@ -98,6 +136,36 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            for word, value in sorted_dict.items():
+                probability += value
+                possible_tokens += (word,)
+                if probability >= self._p_value:
+                    break
+            encoded += (random.choice(possible_tokens),)
+
+        decoded = self._word_processor.decode(encoded)
+        if not decoded:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:166]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 9]
```diff
@@ -125,6 +196,12 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        types = {
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:199]: move to init

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 7 above)
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py:149]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 8 above)
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 9 above)
Comment (mmarina2004) [reply lab_4_fill_words_by_ngrams/main.py]: done

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**